File: 백준/Gold/25682. 체스판 다시 칠하기 2/체스판 다시 칠하기 2.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {

	public static void main(String[] args) throws IOException {

		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine());

		int n = Integer.parseInt(st.nextToken());
		int m = Integer.parseInt(st.nextToken());
		int k = Integer.parseInt(st.nextToken());

		char[][] board = new char[n][m];
		for (int i = 0; i < n; i++) {
			board[i] = br.readLine().toCharArray();
		}

		int[][] sum = new int[n + 1][m + 1];

		for (int i = 1; i <= n; i++) {
			for (int j = 1; j <= m; j++) {
				sum[i][j] = sum[i - 1][j] + sum[i][j - 1] - sum[i - 1][j - 1];
				if ((i + j) % 2 == 0) {
					if (board[i - 1][j - 1] == 'B') {
						sum[i][j]++;
					}
				} else {
					if (board[i - 1][j - 1] == 'W') {
						sum[i][j]++;
					}
				}
			}
		}

		int result = n * m;

		for (int i = k; i <= n; i++) {
			for (int j = k; j <= m; j++) {
				int black = sum[i][j] - sum[i - k][j] - sum[i][j - k] + sum[i - k][j - k];
				int white = k * k - black;
				result = Math.min(result, Math.min(black, white));
			}
		}

		System.out.println(result);
	}
}

File: 백준/Gold/25682. 체스판 다시 칠하기 2/체스판 다시 칠하기 2.py
import sys

n, m, k = map(int, sys.stdin.readline().rstrip().split())
board = []

for _ in range(n):
    board.append(list(sys.stdin.readline().rstrip()))

lst = [[0] * (m + 1) for _ in range(n + 1)]

for i in range(1, n + 1):
    for j in range(1, m + 1):
        lst[i][j] = lst[i - 1][j] + lst[i][j - 1] - lst[i - 1][j - 1]
        if (i + j) % 2 == 0 and board[i - 1][j - 1] == 'B':
            lst[i][j] += 1
        elif (i + j) % 2 == 1 and board[i - 1][j - 1] == 'W':
            lst[i][j] += 1

result = n * m

for i in range(k, n + 1):
    for j in range(k, m + 1):
        black = lst[i][j] - lst[i - k][j] - lst[i][j - k] + lst[i - k][j - k]
        white = k**2 - black
        result = min(result, black, white)

print(result)
